src/core/mesh/qgsmeshtracerenderer.h

```diff
@@ -267,14 +472,80 @@ class QgsMeshVectorStreamLineRenderer: public QgsMeshVectorRenderer
                                      const QgsRectangle &layerExtent,
                                      double magMax );
 
-
     void draw() override;
 
   private:
     std::unique_ptr<QgsMeshStreamField> mStreamLineField;
     QgsRenderContext &mRendererContext;
 };
 
+#endif //SIP_RUN
+
 ///@endcond
 
+/**
+ * \ingroup core
+ *
+ * A wrapper for QgsMeshParticuleTracesField used to render the particles. Available for Python binding
+ *
+ * \since QGIS 3.12
+ */
+class CORE_EXPORT QgsMeshVectorTraceRenderer
+{
+  public:
+    //!Constructor to use from QgsMeshVectorRenderer
+    QgsMeshVectorTraceRenderer( const QgsTriangularMesh &triangularMesh,
+                                const QgsMeshDataBlock &dataSetVectorValues,
+                                const QgsMeshDataBlock &scalarActiveFaceFlagValues,
+                                bool dataIsOnVertices,
+                                const QgsRenderContext &rendererContext,
+                                const QgsRectangle &layerExtent,
+                                double magMax ) SIP_SKIP;
+
+    //!Constructor to use with Python binding
+    QgsMeshVectorTraceRenderer( QgsMeshLayer *layer, const QgsRenderContext &rendererContext );
+
+    //! Destructor
+    ~QgsMeshVectorTraceRenderer();
+
+    //! seeds particles in the vector fields
+    void seedRandomParticles( int count );
+
+    //! Moves all the particles using frame per second (fps) to calculate the displacement
+    QImage imageRendered();
+
+    //! Sets the number of frames per seconds that will be rendered
+    void setFPS( int FPS );
+
+    //! Sets the max number of pixels that can be go through by the particles in 1 second
+    void setMaxSpeedPixel( int max );
+
+    //! Sets maximum life time of particles in seconds
+    void setParticlesLifeTime( double particleLifeTime );
+
+    //! Sets colors of particle
+    void setParticlesColor( const QColor &c );
+
+    //! Sets particle size
+    void setParticlesSize( double width );
+
+    //! Sets the tail factor, used to adjust the length of the tail. 0 : minimum length, >1 increase the tail
+    void setTailFactor( double fct );
+
+    //! Sets the minimum tail length
+    void setMinimumTailLength( int l );
+
+    //! Sets the visual persistence of the tail
+    void setTailPersitence( double p );
+
+  private:
+    QgsMeshParticleTracesField *mParticleField;
```

Comment:
is it owned by the instance of QgsMeshVectorTraceRenderer? if yes use unique_prt

Reply:
My first implantation use unique_ptr, but SIP didn't like... maybe because SIP needs copiable object ?

Reply:
you do not expose unique_ptr in any function that is sipified, but it could still be member private variable. if SIP complains, sometimes you need to declare destructor in cpp file